velox/functions/prestosql/aggregates/ReservoirSampleAggregate.cpp

```diff
@@ -0,0 +1,910 @@
+  } else {
+    const uint64_t random = folly::Random::rand64(processedCount.value(), rng);
+    if (random < samples->size()) {
+      targetIndex = random;
+    }
+  }
+
+  if (targetIndex >= 0) {
+    if (isNull) {
+      samples->setNull(targetIndex, true);
+    } else {
+      samples->copy(vector, targetIndex, row, 1);
+    }
+  }
+}
+
+void ReservoirSampleAccumulator::addValuesToReservoir(
+    const BaseVector* vector,
+    vector_size_t offset,
+    vector_size_t size,
+    folly::ThreadLocalPRNG& rng) {
+  VELOX_CHECK(isInitialized());
+
+  SelectivityVector rows(maxSampleSize.value(), false);
+  std::vector<vector_size_t> toSource(maxSampleSize.value());
+
+  for (vector_size_t i = 0; i < size; i++) {
+    if (processedCount.value() < maxSampleSize) {
+      toSource[processedCount.value()] = offset + i;
+      rows.setValid(processedCount.value(), true);
+
+      processedCount = processedCount.value_or(0) + 1;
+    } else {
+      processedCount = processedCount.value_or(0) + 1;
+
+      const uint64_t random =
+          folly::Random::rand64(processedCount.value(), rng);
+      if (random < samples->size()) {
+        toSource[random] = offset + i;
+        rows.setValid(random, true);
+      }
+    }
+  }
+  rows.updateBounds();
+
+  samples->copy(vector, rows, toSource.data());
+}
+
+void ReservoirSampleAccumulator::mergeSamples(
+    const ArrayVector& otherSampleArrayVector,
+    vector_size_t otherRow,
+    int64_t otherProcessedCount,
+    int32_t otherMaxSampleSize,
+    folly::ThreadLocalPRNG& rng,
+    memory::MemoryPool* pool) {
+  if (otherSampleArrayVector.isNullAt(otherRow)) {
+    return;
+  }
+  mergeSamples(
+      otherSampleArrayVector.elements(),
+      otherSampleArrayVector.offsetAt(otherRow),
+      otherSampleArrayVector.sizeAt(otherRow),
+      otherProcessedCount,
+      otherMaxSampleSize,
+      rng,
+      pool);
+}
+
+BufferPtr makeShuffledIndex(
+    vector_size_t size,
+    folly::ThreadLocalPRNG& randGen,
+    memory::MemoryPool* pool) {
+  BufferPtr index = allocateIndices(size, pool);
+  vector_size_t* rawIndex = index->asMutable<vector_size_t>();
+  std::iota(rawIndex, rawIndex + size, 0);
+  std::shuffle(rawIndex, rawIndex + size, randGen);
+  return index;
+}
+
+void ReservoirSampleAccumulator::mergeSamples(
+    const VectorPtr& otherSampleVector,
+    vector_size_t otherOffset,
+    vector_size_t otherSize,
+    int64_t otherProcessedCount,
+    int32_t otherMaxSampleSize,
+    folly::ThreadLocalPRNG& rng,
+    memory::MemoryPool* pool) {
+  VELOX_DCHECK_GE(otherProcessedCount, 0);
+  VELOX_DCHECK_GT(otherMaxSampleSize, 0);
+  VELOX_CHECK_NOT_NULL(otherSampleVector);
+
+  const TypePtr& elementType = otherSampleVector->type();
+
+  if (!isInitialized()) {
+    initialize(otherMaxSampleSize, elementType, pool);
+  }
+
+  VELOX_DCHECK(
+      samples->type()->equivalent(*elementType),
+      "Samples to be merged are of different types.");
+  VELOX_CHECK_EQ(
+      maxSampleSize.value(),
```

Comment:
if max sample sizes are supposed to be same across accumulators why pass and check if they are same, isnt it going to be same constant used across the whole aggregate ?